File: app/src/main/java/com/fusoft/walkboner/AddInfluencerActivity.java
package com.fusoft.walkboner;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.bumptech.glide.Glide;
import com.fusoft.walkboner.database.DeleteImage;
import com.fusoft.walkboner.database.ImageUploadListener;
import com.fusoft.walkboner.database.StorageDirectory;
import com.fusoft.walkboner.database.UploadImage;
import com.fusoft.walkboner.utils.UidGenerator;
import com.fusoft.walkboner.views.dialogs.InfoDialog;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.textview.MaterialTextView;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QuerySnapshot;

import java.sql.Timestamp;
import java.util.HashMap;
import java.util.List;

import de.dlyt.yanndroid.oneui.dialog.ProgressDialog;
import de.dlyt.yanndroid.oneui.utils.OnSingleClickListener;
import de.dlyt.yanndroid.oneui.view.Toast;
import de.dlyt.yanndroid.oneui.widget.ProgressBar;

public class AddInfluencerActivity extends AppCompatActivity {
    private LinearLayout userBlockedLinear;
    private MaterialTextView userBlockedReasonText;
    private MaterialTextView pickFromGalleryText;
    private ImageView image;
    private EditText firstNameEdittext;
    private EditText lastNameEdittext;
    private EditText nickNameEdittext;
    private EditText instagramUrlEdittext;
    private EditText tiktokUrlEdittext;
    private EditText youtubeUrlEdittext;
    private MaterialButton doneButton;
    private LinearLayout uploadingImageLinear;
    private ProgressBar uploadingProgressBar;
    private MaterialTextView uploadingStatusText;

    private String uploadedImagePath = "";

    private FirebaseFirestore firestore;

    private ProgressDialog loading;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_influencer);

        initView();
        setup();
    }

    private void initView() {
        userBlockedLinear = (LinearLayout) findViewById(R.id.user_blocked_linear);
        userBlockedReasonText = (MaterialTextView) findViewById(R.id.user_blocked_reason_text);
        pickFromGalleryText = (MaterialTextView) findViewById(R.id.pick_from_gallery_text);
        image = (ImageView) findViewById(R.id.image);
        firstNameEdittext = (EditText) findViewById(R.id.first_name_edittext);
        lastNameEdittext = (EditText) findViewById(R.id.last_name_edittext);
        nickNameEdittext = (EditText) findViewById(R.id.nick_name_edittext);
        instagramUrlEdittext = (EditText) findViewById(R.id.instagram_url_edittext);
        tiktokUrlEdittext = (EditText) findViewById(R.id.tiktok_url_edittext);
        youtubeUrlEdittext = (EditText) findViewById(R.id.youtube_url_edittext);
        doneButton = (MaterialButton) findViewById(R.id.done_button);
        uploadingImageLinear = (LinearLayout) findViewById(R.id.uploading_image_linear);
        uploadingProgressBar = (ProgressBar) findViewById(R.id.uploading_progress_bar);
        uploadingStatusText = (MaterialTextView) findViewById(R.id.uploading_status_text);

        firestore = FirebaseFirestore.getInstance();

        loading = new ProgressDialog(AddInfluencerActivity.this);
        loading.setIndeterminate(true);
        loading.setCancelable(false);
        loading.setCanceledOnTouchOutside(false);
        loading.setProgressStyle(ProgressDialog.STYLE_CIRCLE);
    }

    private void setup() {
        pickFromGalleryText.setOnClickListener(v -> imageChooser());

        image.setOnClickListener(v -> imageChooser());

        doneButton.setOnClickListener(new OnSingleClickListener() {
            @Override
            public void onSingleClick(View view) {
                new InfoDialog().MakeDialog(AddInfluencerActivity.this, "Każde zgłoszenie jest weryfikowane przed ostatecznym dodaniem. Wyraźne naruszanie zasad będzie skutkować okresowym lub permamentnym banem.", "Wyślij", "Anuluj", new InfoDialog.InfoDialogInterfaceListener() {
                    @Override
                    public void OnPositiveButtonClicked() {
                        loading.show();
                        sendInfluToModerators();
                    }

                    @Override
                    public void OnNegativeButtonClicked() {

                    }
                });
            }
        });
    }

    private void sendInfluToModerators() {
        Timestamp timestamp = new Timestamp(System.currentTimeMillis());

        FirebaseAuth auth = FirebaseAuth.getInstance();
        FirebaseUser user = auth.getCurrentUser();

        HashMap<String, Object> map = new HashMap<>();
        map.put("influencerUid", UidGenerator.Generate());
        map.put("influencerAddedBy", user.getUid());
        if (firstNameEdittext.getText().toString().isEmpty()) {
            map.put("influencerFirstName", "");
        } else {
            map.put("influencerFirstName", firstNameEdittext.getText().toString());
        }
        if (lastNameEdittext.getText().toString().isEmpty()) {
            map.put("influencerLastName", "");
        } else {
            map.put("influencerLastName", lastNameEdittext.getText().toString());
        }
        map.put("influencerNickName", nickNameEdittext.getText().toString());
        map.put("influencerDescription", "");
        map.put("influencerAvatar", uploadedImagePath);
        map.put("influencerInstagramLink", instagramUrlEdittext.getText().toString());
        map.put("influencerYouTubeLink", youtubeUrlEdittext.getText().toString());
        map.put("influencerTikTokLink", tiktokUrlEdittext.getText().toString());
        map.put("influencerModeratorUid", "");
        map.put("isVerified", "false");
        map.put("isPremium", "false");
        map.put("isHidden", "false");
        map.put("hasInstagram", !instagramUrlEdittext.getText().toString().isEmpty());
        map.put("hasYouTube", !youtubeUrlEdittext.getText().toString().isEmpty());
        map.put("hasTikTok", !tiktokUrlEdittext.getText().toString().isEmpty());
        map.put("isMaintained", "false");
        map.put("isUserFollowing", "true");
        map.put("influencerAddedAt", timestamp.getTime());
        map.put("maintainedTo", timestamp.getTime());
        map.put("viewsCount", 0);

        firestore.collection("moderation").document("influencers").collection("toModerate").add(map).addOnSuccessListener(documentReference -> {
            loading.dismiss();
            Toast.makeText(AddInfluencerActivity.this, "Pomyślnie wysłano!\nDecyzję dostaniesz w powiadomieniu.", Toast.LENGTH_LONG).show();
            finish();
        }).addOnFailureListener(e -> {

        });
    }

    private void imageChooser() {
        Intent i = new Intent();
        i.setType("image/*");
        i.setAction(Intent.ACTION_GET_CONTENT);

        activityResultLauncher.launch(i);
    }

    ActivityResultLauncher<Intent> activityResultLauncher = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(),
            result -> {
                if (result.getResultCode() == Activity.RESULT_OK) {
                    Uri filePath = result.getData().getData();
                    Glide.with(AddInfluencerActivity.this).load(filePath).into(image);
                    pickFromGalleryText.setVisibility(View.GONE);
                    image.setVisibility(View.VISIBLE);

                    if (!uploadedImagePath.isEmpty()) {
                        DeleteImage.Delete(AddInfluencerActivity.this, uploadedImagePath, new DeleteImage.DeleteTask() {
                            @Override
                            public void OnDeleted() {
                                Toast.makeText(AddInfluencerActivity.this, "Poprzednie zdjęcie zostało usunięte z bazy.", Toast.LENGTH_SHORT).show();
                            }

                            @Override
                            public void OnError(String reason) {

                            }
                        });
                        uploadedImagePath = "";
                    }

                    uploadingImageLinear.setVisibility(View.VISIBLE);
                    new UploadImage(AddInfluencerActivity.this, StorageDirectory.INFLUENCER_PATH, filePath, new ImageUploadListener() {
                        @Override
                        public void OnImageUploaded(String imageUrl) {
                            uploadedImagePath = imageUrl;
                            uploadingImageLinear.setVisibility(View.GONE);
                        }

                        @Override
                        public void Progress(int value) {
                            uploadingProgressBar.setProgress(value);
                        }

                        @Override
                        public void OnError(String reason) {
                            uploadingProgressBar.setVisibility(View.GONE);
                            uploadingStatusText.setText(reason);
                        }
                    });
                }
            });
}

File: app/src/main/java/com/fusoft/walkboner/AddInfluencerActivity.kt
/*
 * Copyright (c) 2022 - WalkBoner.
 * AddInfluencerActivity.kt
 * created at: 06.11.2022, 20:16
 * last modified: 06.11.2022, 21:48
 */

package com.fusoft.walkboner

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.result.ActivityResultCallback
import androidx.activity.result.contract.ActivityResultContract
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import com.bumptech.glide.Glide
import com.fusoft.walkboner.adapters.spinner.InfluencersNicksAdapter
import com.fusoft.walkboner.database.DeleteImage
import com.fusoft.walkboner.database.DeleteImage.DeleteTask
import com.fusoft.walkboner.database.ImageUploadListener
import com.fusoft.walkboner.database.StorageDirectory
import com.fusoft.walkboner.database.UploadImage
import com.fusoft.walkboner.databinding.ActivityAddInfluencerBinding
import com.fusoft.walkboner.utils.UidGenerator
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.QuerySnapshot
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.tasks.await
import java.sql.Timestamp
import com.fusoft.walkboner.R.string as AppText

class AddInfluencerActivity : AppCompatActivity() {

    lateinit var binding: ActivityAddInfluencerBinding
    var database: FirebaseFirestore? = null
    var auth: FirebaseAuth? = null

    var uploadedImageUrl = ""

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityAddInfluencerBinding.inflate(layoutInflater)
        setContentView(binding.root)

        init()
        setup()
    }

    override fun onDestroy() {
        database = null
        auth = null

        super.onDestroy()
    }

    @Deprecated("Deprecated in Java")
    override fun onBackPressed() {
        // If image was already uploaded to database, delete.
        if (uploadedImageUrl.isNotEmpty()) {
            binding.loadingView.show()
            DeleteImage.Delete(
                this@AddInfluencerActivity,
                uploadedImageUrl,
                object : DeleteTask {
                    override fun OnDeleted() {
                        binding.loadingView.hide()
                        finish()
                    }

                    override fun OnError(reason: String?) {
                        finish()
                    }
                })
        } else {
            super.onBackPressed()
        }
    }

    private fun init() {
        database = FirebaseFirestore.getInstance()
        auth = FirebaseAuth.getInstance()
    }

    private fun setup() {
        binding.pickFromGalleryText.setOnClickListener {
            imageChooser()
        }

        binding.doneButton.setOnClickListener {
            if (validateInputs()) {
                binding.loadingView.show()
                sendInfluencerToModerators()
            }
        }
    }

    private fun sendInfluencerToModerators() {
        val timestamp = Timestamp(System.currentTimeMillis())

        val user = auth?.currentUser

        val map = HashMap<String, Any>()
        map["influencerUid"] = UidGenerator.Generate()
        map["influencerAddedBy"] = user!!.uid
        if (binding.firstNameEdittext.text.toString().isEmpty()) {
            map["influencerFirstName"] = ""
        } else {
            map["influencerFirstName"] = binding.firstNameEdittext.text.toString().trim()
        }
        if (binding.lastNameEdittext.text.toString().isEmpty()) {
            map["influencerLastName"] = ""
        } else {
            map["influencerLastName"] = binding.lastNameEdittext.text.toString().trim()
        }
        map["influencerNickName"] = binding.nickNameEdittext.text.toString().trim()
        map["influencerDescription"] = ""
        map["influencerAvatar"] = uploadedImageUrl
        map["influencerInstagramLink"] = binding.instagramUrlEdittext.text.toString().trim()
        map["influencerYouTubeLink"] = binding.youtubeUrlEdittext.text.toString().trim()
        map["influencerTikTokLink"] = binding.tiktokUrlEdittext.text.toString().trim()
        map["influencerModeratorUid"] = ""
        map["isVerified"] = "false"
        map["isPremium"] = "false"
        map["isHidden"] = "false"
        map["hasInstagram"] = !binding.instagramUrlEdittext.text.toString().isEmpty()
        map["hasYouTube"] = !binding.youtubeUrlEdittext.text.toString().isEmpty()
        map["hasTikTok"] = !binding.tiktokUrlEdittext.text.toString().isEmpty()
        map["isMaintained"] = "false"
        map["isUserFollowing"] = "true"
        map["influencerAddedAt"] = timestamp.time
        map["maintainedTo"] = timestamp.time
        map["viewsCount"] = 0

        database
            ?.collection("moderation")
            ?.document("influencers")
            ?.collection("toModerate")
            ?.add(map)
            ?.addOnSuccessListener {
                binding.loadingView.hide()
                Toast.makeText(
                    this@AddInfluencerActivity,
                    getString(AppText.influencer_sended_successfuly),
                    Toast.LENGTH_LONG
                ).show()
                finish()
            }?.addOnFailureListener { e: Exception? ->
                Toast.makeText(
                    this@AddInfluencerActivity,
                    getString(AppText.try_again) + e?.message,
                    Toast.LENGTH_LONG
                )
            }
    }

    private fun validateInputs(): Boolean {
        val nickName = binding.nickNameEdittext.text.toString()

        if (nickName.isEmpty()) {
            binding.nickNameEdittext.error = getString(AppText.field_cant_be_empty)
            return false
        } else if (nickName.length >= 16) {
            binding.nickNameEdittext.error = getString(AppText.too_many_chars)
            return false
        } else {
            binding.nickNameEdittext.error = null
            return true
        }
    }

    private fun imageChooser() {
        val i = Intent()
        i.type = "image/*"
        i.action = Intent.ACTION_GET_CONTENT

        activityResultLauncher.launch(i)
    }

    /*
     * FIXME: Wait For Fix from Google
     * Because of Android Studio Bug
     * we need to add this extension to dismiss
     * registerForActivityResult unresolved reference
     * https://stackoverflow.com/a/74139700/13680873
     */
    fun <I, O> Activity.registerForActivityResult(
        contract: ActivityResultContract<I, O>,
        callback: ActivityResultCallback<O>
    ) = (this as ComponentActivity).registerForActivityResult(contract, callback)
    /* This Function does nothing */

    var activityResultLauncher = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result ->
        if (result.resultCode == Activity.RESULT_OK) {
            val filePath = result.data!!.data
            Glide.with(this@AddInfluencerActivity).load(filePath).into(binding.image)
            binding.pickFromGalleryText.visibility = View.GONE
            binding.image.visibility = View.VISIBLE
            if (!uploadedImageUrl.isEmpty()) {
                DeleteImage.Delete(
                    this@AddInfluencerActivity,
                    uploadedImageUrl,
                    object : DeleteTask {
                        override fun OnDeleted() {
                            Toast.makeText(
                                this@AddInfluencerActivity,
                                "Poprzednie zdjęcie zostało usunięte z bazy.",
                                Toast.LENGTH_SHORT
                            ).show()
                        }

                        override fun OnError(reason: String) {}
                    })
                uploadedImageUrl = ""
            }
            binding.uploadingImageLinear.visibility = View.VISIBLE
            UploadImage(
                this@AddInfluencerActivity,
                StorageDirectory.INFLUENCER_PATH,
                filePath,
                object : ImageUploadListener {
                    override fun OnImageUploaded(imageUrl: String) {
                        uploadedImageUrl = imageUrl
                        binding.uploadingImageLinear.visibility = View.GONE
                    }

                    override fun Progress(value: Int) {
                        binding.uploadingProgressBar.progress = value
                    }

                    override fun OnError(reason: String) {
                        binding.uploadingProgressBar.visibility = View.GONE
                        binding.uploadingStatusText.text = reason
                    }
                }
            )
        }
    }
}